core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java

```diff
@@ -0,0 +1,181 @@
+
+        if (includeFrom) {
+            from = rangeType.nextDown(from);
+        }
+        if (includeTo) {
+            to = rangeType.nextUp(to);
+        }
+        expectedRange = rangeType.encodeRange(from, to);
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores, float boost) throws IOException {
+        return new ConstantScoreWeight(this, boost) {
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                final BinaryDocValues values = context.reader().getBinaryDocValues(fieldName);
+                final TwoPhaseIterator iterator = new TwoPhaseIterator(values) {
+
+                    @Override
+                    public boolean matches() throws IOException {
+                        BytesRef encodedRanges = values.binaryValue();
+                        return expectedRange.compare(encodedRanges, queryType);
+                    }
+
+                    @Override
+                    public float matchCost() {
+                        return 4; // at most 4 comparisons
+                    }
+                };
+                return new ConstantScoreScorer(this, score(), iterator);
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "BinaryDocValuesRangeQuery(fieldName=" + field + ")";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BinaryDocValuesRangeQuery that = (BinaryDocValuesRangeQuery) o;
+        return Objects.equals(fieldName, that.fieldName) &&
+                queryType == that.queryType &&
+                Objects.equals(expectedRange, that.expectedRange);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getClass(), fieldName, queryType, expectedRange);
+    }
+
+    /* Range where the includeFrom and includeTo have already been applied */
+    public static final class NormalizedRange {
+
+        private final BytesRef from;
+        private final BytesRef to;
+
+        public NormalizedRange(BytesRef from, BytesRef to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        boolean compare(BytesRef encodedRanges, QueryType queryType) {
+            ByteArrayDataInput in = new ByteArrayDataInput(encodedRanges.bytes, encodedRanges.offset, encodedRanges.length);
+            int numRanges = in.readVInt();
+            for (int i = 0; i < numRanges; i++) {
+                int length = in.readVInt();
+                byte[] bytes = new byte[length];
+                in.readBytes(bytes, 0, length);
+                BytesRef otherFrom = new BytesRef(bytes);
+
+                length = in.readVInt();
+                bytes = new byte[length];
+                in.readBytes(bytes, 0, length);
+                BytesRef otherTo = new BytesRef(bytes);
```

Comment:
I can remove the ByteRef instance creation and use just plain byte[]. 
The reason I used `ByteArrayDataInput` is because of its `readVInt()` and `readBytes()` methods. Are there static alternatives that I can use? Otherwise maybe it is sufficient enough if I keep a spare instance around that I keep reusing?